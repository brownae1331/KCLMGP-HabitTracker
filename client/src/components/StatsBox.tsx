import React from "react";
import { View, Text } from "react-native";
import { Svg, Circle } from "react-native-svg";
import { ThemedText } from "./ThemedText";
import { CalendarPageStyles } from "./styles/CalendarPageStyles";
import { Colors } from "./styles/Colors";
import { useTheme } from "./ThemeContext";

interface StatsBoxComponentProps {
    selectedDate: string;
    completionPercentage: number;
    formatDate: (date: string) => string;
}

export const StatsBoxComponent: React.FC<StatsBoxComponentProps> = ({
    selectedDate,
    completionPercentage,
    formatDate,
}) => {
    const { theme } = useTheme();
    const today = new Date().toISOString().split("T")[0];

    return (
        <View
            style={[
                CalendarPageStyles.statsContainer,
                { backgroundColor: theme === "dark" ? "#1E1E1E" : "#FFFFFF" },
            ]}
        >
            <ThemedText type="subtitle">
                üî• Current Streak: <Text style={{ color: "#FFD700" }}>{17} days</Text>
            </ThemedText>
            <ThemedText type="subtitle">
                üèÜ Longest Streak: <Text style={{ color: "#FFD700" }}>{38} days</Text>
            </ThemedText>

            {/* Separating Line */}
            <View
                style={[
                    CalendarPageStyles.separator,
                    { backgroundColor: theme === "dark" ? "#444444" : "#CCCCCC" },
                ]}
            />

            {/* Habits Completed Header */}
            <View style={CalendarPageStyles.habitsContainer}>
                <ThemedText style={{ fontSize: 24, fontWeight: "bold" }}>
                    Habits completed{" "}
                    <Text style={{ color: "#FFD700" }}>
                        {selectedDate === today ? "Today" : formatDate(selectedDate)}
                    </Text>
                </ThemedText>
            </View>

            {/* Large Progress Circle */}
            <View style={CalendarPageStyles.progressContainer}>
                <Svg width={120} height={120} viewBox="0 0 120 120">
                    <Circle
                        cx="60"
                        cy="60"
                        r="50"
                        stroke={theme === "dark" ? "#444444" : "#CCCCCC"}
                        strokeWidth="10"
                        fill="none"
                    />
                    <Circle
                        cx="60"
                        cy="60"
                        r="50"
                        stroke="#FFD700"
                        strokeWidth="10"
                        fill="none"
                        strokeDasharray={2 * Math.PI * 50}
                        strokeDashoffset={(1 - completionPercentage / 100) * 2 * Math.PI * 50}
                        strokeLinecap="round"
                    />
                </Svg>
                <View style={CalendarPageStyles.progressTextContainer}>
                    <ThemedText
                        style={[CalendarPageStyles.percentageText, { color: "#FFD700" }]}
                    >
                        {completionPercentage}%
                    </ThemedText>
                    <ThemedText style={CalendarPageStyles.fractionText}>
                        {completionPercentage}/100
                    </ThemedText>
                </View>
            </View>
        </View>
    );
}; 