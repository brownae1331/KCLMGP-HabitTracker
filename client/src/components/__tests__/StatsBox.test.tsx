import React from 'react';
import { render } from '@testing-library/react-native';
import { StatsBoxComponent } from '../StatsBox';

// Mock ThemeContext
jest.mock('../ThemeContext', () => ({
    useTheme: () => ({ theme: 'light' }),
}));

describe('StatsBoxComponent', () => {
    it('displays current and longest streak values correctly', () => {
        const { getByText } = render(
            <StatsBoxComponent
                selectedDate="2025-03-25"
                completionPercentage={75}
                formatDate={(date) => date}
                currentStreak={5}
                longestStreak={10}
            />
        );
        // Verify streak texts
        expect(getByText('5 days')).toBeTruthy();
        expect(getByText('10 days')).toBeTruthy();
    });

    it('displays the completion percentage and circle correctly', () => {
        const { getByText, getByText: queryByText } = render(
            <StatsBoxComponent
                selectedDate="2025-03-25"
                completionPercentage={100}
                formatDate={(date) => date}
                currentStreak={0}
                longestStreak={0}
            />
        );
        // 100% should be displayed and the circle strokeDashoffset should be 0
        expect(getByText('100%')).toBeTruthy();
        // The 0% case (if needed) would show full offset; test 100% for zero offset
        const componentTree = queryByText('100%'); // using the 100% text node to traverse up
    });
});
