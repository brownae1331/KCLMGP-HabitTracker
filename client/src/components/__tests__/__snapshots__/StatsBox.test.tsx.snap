// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsBoxes Component matches snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-between",
        "marginTop": "5%",
        "paddingBottom": "5%",
        "paddingHorizontal": "5%",
      },
      {
        "backgroundColor": "#f0f0f0",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "marginBottom": "5%",
          "padding": 15,
          "width": "45%",
        },
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 24,
              "fontWeight": "bold",
            },
            {
              "color": "#11181C",
            },
          ],
        ]
      }
    >
      5
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 14,
              "marginTop": 4,
            },
            {
              "color": "#AFAFAFFF",
            },
          ],
        ]
      }
    >
      Current Streak
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "marginBottom": "5%",
          "padding": 15,
          "width": "45%",
        },
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 24,
              "fontWeight": "bold",
            },
            {
              "color": "#11181C",
            },
          ],
        ]
      }
    >
      10
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 14,
              "marginTop": 4,
            },
            {
              "color": "#AFAFAFFF",
            },
          ],
        ]
      }
    >
      Longest Streak
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "marginBottom": "5%",
          "padding": 15,
          "width": "45%",
        },
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 24,
              "fontWeight": "bold",
            },
            {
              "color": "#11181C",
            },
          ],
        ]
      }
    >
      80%
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 14,
              "marginTop": 4,
            },
            {
              "color": "#AFAFAFFF",
            },
          ],
        ]
      }
    >
      completion Rate
    </Text>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "borderWidth": 1,
          "marginBottom": "5%",
          "padding": 15,
          "width": "45%",
        },
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 24,
              "fontWeight": "bold",
            },
            {
              "color": "#11181C",
            },
          ],
        ]
      }
    >
      Test Value
    </Text>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "fontSize": 14,
              "marginTop": 4,
            },
            {
              "color": "#AFAFAFFF",
            },
          ],
        ]
      }
    >
      Test Stat
    </Text>
  </View>
</View>
`;
